crates/datasources/src/json/stream.rs

```diff
@@ -0,0 +1,83 @@
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+use datafusion::arrow::datatypes::{Schema, SchemaRef};
+use datafusion::arrow::json::ReaderBuilder;
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::error::DataFusionError;
+use datafusion::execution::TaskContext;
+use datafusion::physical_plan::streaming::PartitionStream;
+use datafusion::physical_plan::{RecordBatchStream, SendableRecordBatchStream};
+use futures::{Stream, StreamExt};
+use serde_json::{Map, Value};
+
+type SendableCheckedRecordBatchStrem =
+    Pin<Box<dyn Stream<Item = Result<RecordBatch, DataFusionError>> + Send>>;
+
+pub struct JsonStream {
+    schema: Arc<Schema>,
+    stream: SendableCheckedRecordBatchStrem,
+}
+
+impl Stream for JsonStream {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.stream.poll_next_unpin(cx)
+    }
+}
+
+impl RecordBatchStream for JsonStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+pub struct JsonPartitionStream {
+    schema: Arc<Schema>,
+    stream: Mutex<Option<SendableCheckedRecordBatchStrem>>,
+}
+
+impl PartitionStream for JsonPartitionStream {
+    fn schema(&self) -> &SchemaRef {
+        &self.schema
+    }
+
+    fn execute(&self, _ctx: Arc<TaskContext>) -> SendableRecordBatchStream {
+        let partition = self
+            .stream
+            .lock()
+            .unwrap()
+            .take()
+            .expect("stream to only be called once")
+            .boxed();
+
+        Box::pin(JsonStream {
+            schema: self.schema.clone(),
+            stream: partition,
+        })
+    }
+}
+
+impl JsonPartitionStream {
+    pub fn new(schema: Arc<Schema>, chunk: Vec<Map<String, Value>>) -> Self {
+        let stream_schema = schema.clone();
+        let stream = futures::stream::iter(chunk)
+            .chunks(25)
+            .map(move |objs| {
+                let mut decoder =
+                    ReaderBuilder::new(stream_schema.clone().to_owned()).build_decoder()?;
+                decoder
+                    .serialize(&objs)
+                    .map_err(|e| DataFusionError::External(Box::new(e)))?;
+                Ok(decoder.flush()?.unwrap())
```

Comment:
this seems really inefficient. We first deserialize the json buffer to a `serde_json::Value`, then we re-serialize just to deserialize it again as a RecordBatch?

so it goes from `&[u8]` -> `Vec<Map<String, Value>>` -> `arrow_json::Tape` -> `RecordBatch`

Reply:
you are correct.

you can't make a record batch or or an arrow-json decoder without a schema, and you need to examine the document's keys/values to get the schema. 

on the plus side, it's all in memory and the amount of time you spend fiddling with the rust abstractions is definitely a fraction of the time you spend reading the data off the filesystem or out of object_store/http.